libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -531,10 +531,14 @@ void CompilerUtils::encodeToMemory(
 			if (_givenTypes[i]->category() == Type::Category::StringLiteral)
 			{
 				auto const& strType = dynamic_cast<StringLiteralType const&>(*_givenTypes[i]);
-				m_context << u256(strType.value().size());
+				auto const size = strType.value().size();
+				m_context << u256(size);
 				storeInMemoryDynamic(*TypeProvider::uint256(), true);
 				// stack: ... <end_of_mem'>
-				storeInMemoryDynamic(strType, _padToWordBoundaries);
+
+				// Do not output empty padding for zero-length strings.
+				if (size != 0)
+					storeInMemoryDynamic(strType, _padToWordBoundaries);
```

Comment:
Why not modify `storeInMemoryDynamic` instead? How does `string memory x = ""; revert(x);` behave?

Reply:
> How does string memory x = ""; revert(x); behave?

Apparently no padding in that case. See #10193.

Reply:
So the only affected type was `StringLiteral`?
If so, I think this is ok.

Reply:
If I move this check to `storeInMemoryDynamic` then this fails:
```
semanticTests/various/literal_empty_string.sol: FAIL
  Contract:
    contract C {
        bytes32 public x;
        uint256 public a;
    
        function f(bytes32 _x, uint256 _a) public {
            x = _x;
            a = _a;
        }
    
        function g() public {
            this.f("", 2);
        }
    }
    
    // ====
    // compileViaYul: also

  Expected result:
  // x() -> 0
  // a() -> 0
  // g() ->
  // x() -> 0
  // a() -> 2

  Obtained result:
  // x() -> 0
  // a() -> 0
  // g() -> FAILURE
  Warning: The call to "g()" returned 

  // x() -> 0
  // a() -> 0
  Warning: The call to "a()" returned 
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
  
  Attention: Updates on the test will apply the detected format displayed.
  
  Note that the test also has to pass via Yul.

(e)dit/(u)pdate expectations/(s)kip/(q)uit? 
```

Reply:
@chriseth this has been blocked on the above, I still didn't had the patience to debug this. Checked all` storeInMemoryDynamic` calls roughly, and none of them seemed suspect.

At least we should merge the comment improvements. We could merge this fix/improvement, and we could also try to find the reason -- however I'm not sure we need/have to do this.

Reply:
Moved this into an issue.